### PR TITLE
Add Serialization library

### PR DESCRIPTION
**Pull request by knelli2**
This moves some files from the Parallel library that only have to do with serialization to this new Serialization library to avoid unwanteded dependencies on Parallel

### Upgrade instructions

If you have one of the following files
- src/Parallel/PupStlCpp11.hpp
- src/Parallel/PupStlCpp17.hpp 
- src/Parallel/CharmPupable.hpp
- src/Parallel/RegisterDerivedClassesWithCharm.hpp
- src/Parallel/PupBoost.hpp
- src/Parallel/Serialize.hpp

they are now in `src/Utilities/Serialization/`. Also, all functions in `RegisterDerivedClassesWithCharm.hpp` were moved from the Parallel namespace to the global namespace.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Utilities/CMakeLists.txt [hunk 1]
```diff
@@ -92,6 +92,7 @@ target_link_libraries(
 
 add_subdirectory(ErrorHandling)
 add_subdirectory(System)
+add_subdirectory(Serialization)
```
Comment (nilsvu) [inline src/Utilities/CMakeLists.txt:95]: (nitpick) move up one line for alphabetization

File: src/Utilities/Serialization/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+set(LIBRARY Serialization)
+
+add_spectre_library(${LIBRARY} INTERFACE)
+
+# spectre_target_sources(
```
Comment (nilsvu) [inline src/Utilities/Serialization/CMakeLists.txt:8]: delete?

File: src/Utilities/Serialization/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  PupStlCpp17.hpp
+  RegisterDerivedClassesWithCharm.hpp
+  Serialize.hpp
+  )
+
+target_link_libraries(
+  ${LIBRARY}
+  INTERFACE
+  Charmxx::pup
+  Utilities
```
Comment (nilsvu) [inline src/Utilities/Serialization/CMakeLists.txt:29]: add `Boost::boost`

File: tests/Unit/Utilities/CMakeLists.txt [hunk 4]
```diff
@@ -51,10 +51,11 @@ set(LIBRARY_SOURCES
...
 
 add_subdirectory(TypeTraits)
+add_subdirectory(Serialization)
 
 add_test_library(
   ${LIBRARY}
   "Utilities"
   "${LIBRARY_SOURCES}"
-  "Boost::boost;DataStructures;Options;Utilities"
+  "Boost::boost;DataStructures;Options;Serialization;Utilities"
```
Comment (nilsvu) [inline tests/Unit/Utilities/CMakeLists.txt:60]: if you get the chance, do this:

```cmake
add_test_library(
   ${LIBRARY}
   "Utilities"
   "${LIBRARY_SOURCES}"
   ""
  )

target_link_library(
  ${LIBRARY}
  PRIVATE
  ...
  )
```

**Review by nilsvu — APPROVED** [review]
Feel free to fix these things in a follow-up so we can merge this quickly

**Comment by knelli2** [comment]
Yeah I think I'll do those in a follow up

File: tests/Unit/PointwiseFunctions/Hydro/EquationsOfState/Test_SpectralEoS.cpp [hunk 5]
```diff
@@ -10,11 +10,11 @@
 #include "DataStructures/Tensor/Tensor.hpp"
 #include "Framework/TestCreation.hpp"
 #include "Helpers/PointwiseFunctions/Hydro/EquationsOfState/TestHelpers.hpp"
-#include "Parallel/RegisterDerivedClassesWithCharm.hpp"
 #include "PointwiseFunctions/Hydro/EquationsOfState/EquationOfState.hpp"
 #include "PointwiseFunctions/Hydro/EquationsOfState/Factory.hpp"
 #include "PointwiseFunctions/Hydro/EquationsOfState/Spectral.hpp"
 #include "PointwiseFunctions/Hydro/SpecificEnthalpy.hpp"
+#include "Utilities/Serialization/RegisterDerivedClassesWithCharm.hpp"
```
Comment (wthrowe) [inline tests/Unit/PointwiseFunctions/Hydro/EquationsOfState/Test_SpectralEoS.cpp:17]: This is unused, but that's a preexisting issue and I won't make you fix it.

**Review by wthrowe — APPROVED** [review]
clang-tidy and test failures are unrelated.

**Pull request merged**